### PR TITLE
Task/use etcd metrics endpoint

### LINKED ISSUE #9093: Support metrics endpoint in etcd metricbeat module
Current etcd [monitoring documentation](https://github.com/etcd-io/etcd/blob/v3.3.10/Documentation/op-guide/monitoring.md) only mentions the `/metrics` endpoint (prometheus) to collect metrics. Metricbeat module uses the [v2 stats endpoint](https://github.com/etcd-io/etcd/blob/v3.3.10/Documentation/v2/api.md#statistics).

They seem to offer complementary metrics (there are some metrics that I couldn't find in `/metrics` endpoint) but the main problem I see is that v2 API doesn't seem to cover v3 store and gRPC operations, and the `/metrics` endpoint seems to cover both stores and gRPC.

[etcd 3.3](https://coreos.com/blog/announcing-etcd-3.3) includes an experimental v2 emulation layer over v3 stores, we could try to collect v3 stats using this feature, but this would require users to enable non-default experimental features, what is not ideal.

### PR DESCRIPTION
**Pull request by odacremolbap**
Add etcd metrics endpoint for Etcd V3 as a new metricset

Fixes elastic/beats#9093

### CODE REVIEW

**Comment by odacremolbap** [comment]
I need to wrap my head around the unit tests and docs.
Opening the PR in order to get some help while I read upon it at other modules

**Comment by ruflin** [comment]
@odacremolbap Happy to help, let me know where :-)

**Comment by odacremolbap** [comment]
Received some out-of-band feedback on metricset naming:

- Etcd is a single binary that exposes V2 and V3
- Clients choose which one to use, it is expected that since V3 was released, all clients use V3
- Each version (V2/V3) keeps most endpoints and storage separate. No data saved as Vx can be read from Vy.
- Monitoring is also kept separate for each version, V2 will use current beats metricsets (`self`, `leader`, `store`) meanwhile V3 will expose a prometheus formatted `metrics` endpoint. Metrics read for each version will be specific to its version, although I have some doubts regarding memory and disk.
- It is not easy to discover which version is being used. All endpoints are up.

From a user perspective:
- An admin knows what version is being used, it is expected to be V3
- Although it is possible to use V2 and V3 at the same time, that's an anti-pattern to be avoided

At this moment, we are exposing

- `self`
- `store`
- `disk`
- `metrics`

3 first for V2, last one for V3.
We need to come up with a solution for making clear which one to use.

File: metricbeat/module/etcd/metrics/_meta/testdata/metrics.plain [hunk 1]
```diff
@@ -0,0 +1,1256 @@
...
+grpc_server_handled_total{grpc_code="Internal",grpc_method="Status",grpc_service="etcdserverpb.Maintenance",grpc_type="unary"} 0
+grpc_server_handled_total{grpc_code="Internal",grpc_method="Txn",grpc_service="etcdserverpb.KV",grpc_type="unary"} 0
+grpc_server_handled_total{grpc_code="Internal",grpc_method="UserAdd",grpc_service="etcdserverpb.Auth",grpc_type="unary"} 0
+grpc_server_handled_total{grpc_code="Internal",grpc_method="UserChangePassword",grpc_service="etcdserverpb.Auth",grpc_type="unary"} 0
+grpc_server_handled_total{grpc_code="Internal",grpc_method="UserDelete",grpc_service="etcdserverpb.Auth",grpc_type="unary"} 0
+grpc_server_handled_total{grpc_code="Internal",grpc_method="UserGet",grpc_service="etcdserverpb.Auth",grpc_type="unary"} 0
+grpc_server_handled_total{grpc_code="Internal",grpc_method="UserGrantRole",grpc_service="etcdserverpb.Auth",grpc_type="unary"} 0
+grpc_server_handled_total{grpc_code="Internal",grpc_method="UserList",grpc_service="etcdserverpb.Auth",grpc_type="unary"} 0
+grpc_server_handled_total{grpc_code="Internal",grpc_method="UserRevokeRole",grpc_service="etcdserverpb.Auth",grpc_type="unary"} 0
+grpc_server_handled_total{grp
```
Comment (jsoriano) [inline metricbeat/module/etcd/metrics/_meta/testdata/metrics.plain]: It'd be probably nice to expose some of these metrics on methods, but lets leave it for future changes.

File: metricbeat/module/etcd/metrics/_meta/fields.yml [hunk 2]
```diff
@@ -0,0 +1,96 @@
+- name: metrics
+  type: group
```
Comment (jsoriano) [inline metricbeat/module/etcd/metrics/_meta/fields.yml]: Add here `release: beta` to mark this as beta.

**Comment by jsoriano** [comment]
Regarding v2 vs v3 storages. I think we should continue enabling the old metricsets by default, and at some moment we would enable also this new one so we have all covered by default.

I saw that some operations on v2 storage also affect the metrics exposed by the new endpoint, maybe at some moment the new endpoint covers both storages, not sure about their plan on this. If that is the case, maybe we can deprecate the old metricsets at some point, but not before metricbeat 8.

File: metricbeat/module/etcd/metrics/metrics_integration_test.go [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+	je, e := json.Marshal(events)
+	t.Logf("pre metrics: %v", events)
+	r := json.Indent(&j, je, "", "\t")
+	if e != nil {
+		t.Logf("error: %v", e)
+	}
+
+	t.Logf("metrics: %v", r)
+
+}
```
[the remark this answers is not in the document]
Comment (odacremolbap) [reply metricbeat/module/etcd/metrics/metrics_integration_test.go]: Copied from other similar tests, I also wondered and _copied just in case_.
I'll remove

**Comment by ruflin** [comment]
I think there are 2 users: Admin that deploys Metricbeat and he should know if we uses v2 or v3. From a data consumer perspective it should not matter which one is used. My understanding so far is that v3 is mostly a superset of v2. So if a user upgrades from v2 to v3, the resulting events should still look the same but have more data inside.

The above assume that we like the current data structure. If we don't I'm also ok with introducing a new better data structure for v3.

**Comment by odacremolbap** [comment]
@jsoriano Agree

I don't think we have to deprecate V2, it is not deprecated by RedHat/CoreOS afaik, besides current fetch is working nicely.

**Comment by odacremolbap** [comment]
@ruflin 
most V3 metrics could be mapped to V2, but:
- some V2 metrics would appear as empty since the migration
- network metrics that were referring to HTTP now would be gRPC
- V3 did an internal reorganization of metrics, we would be sticking to the previous V2 structure, which might feel artificial and hard to cope with in the future

Also, migrating V2 to V3 is not something you usually do, you need all clients to migrate also. The set of client libraries and their methods changed, so V2 and V3 requires heavily modifying client's codebase.
This would be like migrating from MySql to Postgres and keeping the same metricset, that's possible since both are DBs and metric schema should be similar, but creating an abstraction layer that unifies that schema probably leads to a rigid schema

There is another reason for not translating V3 (and this would be the main one for me), if we translate metrics too far from what they are named from its source, troubleshooting when things go wrong can be hard, since googling won't probably show any result for the new metrics names

**Comment by odacremolbap** [comment]
Regarding module/metricsets:

current schema is
- module `etcd`
- metricsets `store`, `self`, `leader`, `metrics`

The main thing to improve is noticing users which etcd version matches which metrics. Current V2 metricsets (`store`, `self`, `leader`) are GA and being used, imho we should keep it as is for now. `metrics` metricset is V3. The name follows the endpoint where metrics are retrieved, which is not a descriptive name from the user POV.

Choices:
1. keep metricset as they are
2. change `metrics`to something like `metricsV3`
3. refactor V2 metricsets so they become `storeV2`, `selfV2`, `leaderV2` while keeping current names also for backwards compatibility
4. create `etcdV3` as a new module separate from current `etcd``
5. <write here your proposal>

My preferred would be 2
Comments @ruflin @jsoriano ?

File: metricbeat/module/etcd/metrics/_meta/fields.yml [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+    - name: netowrk.client_grpc_received_bytes_total
+      type: long
+      description: >
+        gRPC received bytes total
+
+    - name: netowrk.client_grpc_sent_bytes_total
+      type: long
+      description: >
+        gRPC received sent total
+
```
[the remark this answers is not in the document]
Comment (odacremolbap) [reply metricbeat/module/etcd/metrics/_meta/fields.yml]: thanks, those resources are really handy

**Comment by ruflin** [comment]
As mentioned before, I prefer not to mention V3 in the final doc as for the consumer it should not matter, so just `metrics` should be fine. What I don't like about the `metrics` prefix is that it's a bit meaningless as everything here is metrics.

One other idea triggered by your comment that it's a huge migration from A to B and it's unlikely that both will be used at the same time. What happens if not use the `metrics` prefix and put all the metrics directly under `etcd`? Will it conflict with v2? It's not a typical thing we do for metricsets but if I understand this change here to prometheus correctly this will be the only endpoint available and it's unlikely more endpoints will be added in v3? It's just that more metrics will be added? Based on your example event we would have `etcd.disk.*` etc. an none of these seems to conflict with the current one?

**Comment by odacremolbap** [comment]
generally speaking adding version to metric names is not a good idea, I agree with you, no point on creating metricsets for fooV1 and a new set for fooV2. The problem with etcd is that they sort of bundle 2 products in 1 binary.

As an example, here is how you use the client for V2 for **any** etcd release:
```
etcdctl get myKey
```
and here is how you use V3
```
ETCDCTL_API=3 etcdctl get myKey
```
Admins, devs, users, anyone dealing with etcd must know beforehand if they are targeting V2 or V3. Internally the command will be redirected to a different set of functions based on the environment variable.

So, if i understood it correctly, that proposal would be, keeping current V2
- `etcd.store`
- `etcd.self`
- `etcd.leader`
And add new V3 metrics without a V3 reference at that same level
- `etcd.memory`
- `etcd.network`
- `etcd.server`
- `etcd.disk`

I'm ok with that since it avoids using extremely generic term `metrics`, just wondering if we can come up with something better to highlight that whenever you use V3 you should be using those new ones. Just like `etcdctl` requiring `ETCDCTL_API=3` environment variable.

We might not add any V3 reference to metrics, but as an etcd user in my past life, I would be confused to see `etcd.server` and `etcd.self` without knowing which one is pre-V3 and which one post-V3

If you feel that our best move is setting those names, and clarifying at docs, I'm ok with that, I guess there is no perfect solution since this problem is upstream etcd.

File: metricbeat/module/etcd/metrics/_meta/fields.yml [hunk 5]
```diff
@@ -0,0 +1,96 @@
+- name: metrics
```
[the remark this answers is not in the document]
Comment (odacremolbap) [reply metricbeat/module/etcd/metrics/_meta/fields.yml]: I dont follow

Is this related to the metricset/metrics renaming thread?

**Comment by odacremolbap** [comment]
As discussed out of band with @ruflin 

We will add a field to metricsets that indicate the API version used when retrieving the metrics.
All V2 metrics will have `apiVersion: 2`
All V3 metrics will have `apiVersion: 3`

Although users will still need to get to the docs to check which metricset applies to what apiVersion, this solution has a number of advantages:
- if some V2 metrics are still use for V3 users, they will be around
- still V2 and V3 will be distinguishable and available for filtering at ES
- current V2 users won't be affected

@jsoriano feedback?

File: metricbeat/module/etcd/metrics/_meta/fields.yml (hunk 5 above)
[the remark this answers is not in the document]
Comment (jsoriano) [reply metricbeat/module/etcd/metrics/_meta/fields.yml]: Yes, it'd be for example to use `etcd.server.has_leader` instead of `etcd.metrics.server.has_leader`.

**Comment by jsoriano** [comment]
Ok to add an `apiVersion` field, but taking into account that "v3" metrics also contain metrics for v2 store and endpoint.

**Comment by odacremolbap** [comment]
@jsoriano @ruflin 
I've pushed
- adding apiVersion to both V2 and V3 etcd metrics
- changing the namespace for V3 metrics so that they are consistent with V2 metrics placement

`prometheus.MetricsMapping` needed to be added the `Namespace` field in order for this to work.

I haven't included the updated JSON until we sort out why `agent` group field is missing

File: metricbeat/module/etcd/metrics/metrics.go [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+				prometheus.OpMultiplyBuckets(1000000000)),
+
+			// Memory
+			"go_memstats_alloc_bytes": prometheus.Metric("memory.go_memstats_alloc.bytes"),
+
+			// Network
+			"etcd_network_client_grpc_sent_bytes_total":     prometheus.Metric("network.client_grpc_sent.bytes"),
+			"etcd_network_client_grpc_received_bytes_total": prometheus.Metric("network.client_grpc_received.bytes"),
+		},
+		ExtraFields: map[string]string{"apiVersion": "3"},
```
Comment (jsoriano) [inline metricbeat/module/etcd/metrics/metrics.go]: Please don't use camel case for field names.
```suggestion
		ExtraFields: map[string]string{"api_version": "3"},
```

File: metricbeat/helper/prometheus/prometheus.go [hunk 7]
```diff
@@ -213,7 +216,10 @@ func (p *prometheus) ReportProcessedMetrics(mapping *MetricsMapping, r mb.Report
 		return
 	}
 	for _, event := range events {
-		r.Event(mb.Event{MetricSetFields: event})
+		r.Event(mb.Event{
+			MetricSetFields: event,
+			Namespace:       mapping.Namespace,
+		})
```
Comment (jsoriano) [inline metricbeat/helper/prometheus/prometheus.go]: Nice. I think this should be moved to its own PR, with a note in the developers changelog.

Comment (ruflin) [reply metricbeat/helper/prometheus/prometheus.go]: +1 on having a separate PR. I think @sayden will also be happy to see this.

Comment (odacremolbap) [reply metricbeat/helper/prometheus/prometheus.go]: thanks, 
created https://github.com/elastic/beats/issues/11423
pushed https://github.com/elastic/beats/pull/11424

File: metricbeat/module/etcd/metrics/metrics_integration_test.go [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+	events, errs := mbtest.ReportingFetchV2(f)
+	if len(errs) > 0 {
+		t.Fatalf("Expected 0 error, had %d. %v\n", len(errs), errs)
+	}
+	assert.NotEmpty(t, events)
+	t.Logf("%s/%s event: %+v", f.Module().Name(), f.Name(), events[0])
+}
+
+func TestData(t *testing.T) {
+	compose.EnsureUp(t, "etcd")
```
Comment (ruflin) [inline metricbeat/module/etcd/metrics/metrics_integration_test.go]: As we already have the new test setup for this, we don't need this method I think.
